### PR TITLE
Bug fix: Cloud Functions Runtime Config Hash Value

### PR DESCRIPTION
**Pull request by mercury2269**
### Description

We've been trying to figure out why in our project functions are re-deployed every time, even when the source code and function configuration doesn't change. As we found out, the runtime configuration produces new hash every time. Because during the plan step the retrieved runtime configuration is not sorted, every time it is pulled it's produces different order of configuration keys. And when the configuration is appended to hash string, it produces a new hash every time.   

Solution: 
Convert runtime configuration object tree into a sorted array.

Please see the following reported issue: 
https://github.com/firebase/firebase-tools/issues/5354

### Scenarios Tested

Re-deploying functions previously deployed with the same code base and configuration.  

cc @TheIronDev

### CODE REVIEW

File: src/deploy/functions/prepareFunctionsUpload.ts [hunk 1]
```diff
@@ -88,7 +90,8 @@ async function packageSource(
       });
     }
     if (typeof runtimeConfig !== "undefined") {
-      const runtimeConfigString = JSON.stringify(runtimeConfig, null, 2);
```
Comment (TheIronDev) [inline src/deploy/functions/prepareFunctionsUpload.ts:91]: There is a chance mutating the runtimeConfig string may break function upload behavior.
(@taeold / @inlined can keep me honest here.)

To mitigate that, can you use a separate variable to push into the hash?

```
const runtimeConfigString = JSON.stringify(runtimeConfig, null, 2);
      // In order for hash to be consistent, configuration object tree must be sorted by key, only possible with arrays.
 const runtimeConfigHashString = JSON.stringify(convertToSortedKeyValueArray(runtimeConfig));
hashes.push(runtimeConfigHashString);
```

**Comment by TheIronDev** [comment]
@mercury2269 Thank you for opening this pull request and catching this bug!

Overall the code looks good, I made a comment requesting the config that is added to the archiver be left alone.

Would you be able to add a unit test for this change?

Unfortunately the spec file doesn't exist yet. You would need to create: `src/test/deploy/functions/prepareFunctionsUpload.spec.ts`

 You don't have to write a unit test for the entire file, just the parts that you are making changes to.

File: src/deploy/functions/prepareFunctionsUpload.ts (hunk 1 above)
Comment (taeold) [reply src/deploy/functions/prepareFunctionsUpload.ts:91]: IIUC, this code is mutating the value that we pass into the hash, not the underlying runtime config value.

If I understood correctly, that looks safe!

Comment (mercury2269) [reply src/deploy/functions/prepareFunctionsUpload.ts:91]: @TheIronDev @taeold I think it makes sense for the configuration file in the archive to be pretty printed anyways. I went ahead and added a separate formatted string variable. 

Also, as suggested, added a few tests for the sort function. Please have a look @TheIronDev.

**Review by taeold — APPROVED** [review]
LGTM. Do you mind adding an entry to the Changelog.md describing the change?

Thank you for your contribution!

**Comment by mercury2269** [comment]
@taeold no problem, added.

**Comment by alexpchin** [comment]
It would be great to merge this 👍 🥇

**Comment by taeold** [comment]
@mercury2269 Do you mind resolving the merge conflict? I think there is a merge conflict on CHANGELOG.md file.

**Comment by mercury2269** [comment]
> @mercury2269 Do you mind resolving the merge conflict? I think there is a merge conflict on CHANGELOG.md file.

done @taeold

**Comment by pedrofsantoscom** [comment]
are there any updates to when this PR can be merged?

**Review by taeold — APPROVED** [review]

**Comment by edenhermelin** [comment]
hey @taeold @mercury2269 
I updated to the latest version and looks like the deployment still doesn't skip functions that are deployed with the exact same codebase. I'm wondering if I'm facing the same issue or there is something else.
Do you have a suggestion for how can I investigate further? how can I compare the runtime configs between deployments and make sure they are the same?

Thanks!

**Comment by mercury2269** [comment]
> hey @taeold @mercury2269 I updated to the latest version and looks like the deployment still doesn't skip functions that are deployed with the exact same codebase. I'm wondering if I'm facing the same issue or there is something else. Do you have a suggestion for how can I investigate further? how can I compare the runtime configs between deployments and make sure they are the same?
> 
> Thanks!

Hi @edenhermelin,

The caching feature is working for us now, so it may be an unrelated bug. You can follow contribution readme on how to setup and debug locally [here](https://github.com/firebase/firebase-tools/blob/master/CONTRIBUTING.md)

**Pull request merged**